flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java

```diff
@@ -133,14 +136,14 @@ public boolean isAvailable() {
 	 * @param bufferAndBacklog
 	 * 		current buffer and backlog including information about the next buffer
 	 */
-	private boolean isAvailable(BufferAndBacklog bufferAndBacklog) {
+	@Nullable
+	private Buffer.DataType getNextDataType(BufferAndBacklog bufferAndBacklog) {
 		// BEWARE: this must be in sync with #isAvailable()!
-		if (numCreditsAvailable > 0) {
-			return bufferAndBacklog.isDataAvailable();
-		}
-		else {
-			return bufferAndBacklog.isEventAvailable();
+		final Buffer.DataType nextDataType = bufferAndBacklog.getNextDataType();
+		if (numCreditsAvailable > 0 || (nextDataType != null && nextDataType.isEvent())) {
+			return nextDataType;
 		}
+		return null;
```

Comment:
An enum type NONE would work for me and might make the code a bit clearer. However, be aware that this is mostly a copy&replace; I don't think it would simplify any code path.